pkg/apis/networking/v1alpha1/clusteringress_validation.go

```diff
@@ -28,130 +28,127 @@ func (ci *ClusterIngress) Validate() *apis.FieldError {
 	return ci.Spec.Validate().ViaField("spec")
 }
 
-func (spec *ClusterIngressSpec) Validate() *apis.FieldError {
-	fmt.Printf("%#v\n", spec)
+func (spec *IngressSpec) Validate() *apis.FieldError {
 	// Spec must not be empty.
-	if equality.Semantic.DeepEqual(spec, &ClusterIngressSpec{}) {
+	if equality.Semantic.DeepEqual(spec, &IngressSpec{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
+	var all *apis.FieldError = nil
 	// Spec must have at least one rule.
 	if len(spec.Rules) == 0 {
-		return apis.ErrMissingField("rules")
+		all = all.Also(apis.ErrMissingField("rules"))
 	}
 	// Validate each rule.
 	for idx, rule := range spec.Rules {
 		if err := rule.Validate(); err != nil {
-			return err.ViaField(fmt.Sprintf("rules[%d]", idx))
+			all = all.Also(err.ViaField(fmt.Sprintf("rules[%d]", idx)))
 		}
 	}
 	// TLS settings are optional.  However, all provided settings should be valid.
 	for idx, tls := range spec.TLS {
 		if err := tls.Validate(); err != nil {
-			return err.ViaField(fmt.Sprintf("tls[%d]", idx))
+			all = all.Also(err.ViaField(fmt.Sprintf("tls[%d]", idx)))
 		}
 	}
-	return nil
+	return all
 }
 
 func (r *ClusterIngressRule) Validate() *apis.FieldError {
 	// Provided rule must not be empty.
 	if equality.Semantic.DeepEqual(r, &ClusterIngressRule{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
-	if len(r.Hosts) == 0 {
-		return apis.ErrMissingField("hosts")
-	}
-	if err := validateRuleValue(r.ClusterIngressRuleValue); err != nil {
-		return err
-	}
-	return nil
-}
-
-func validateRuleValue(r ClusterIngressRuleValue) *apis.FieldError {
-	// At least one kind of ClusterIngressRuleValue is required.
-	// Since we only have HTTPClusterIngressRuleValue now, it must be provided.
+	var all *apis.FieldError = nil
 	if r.HTTP == nil {
-		return apis.ErrMissingField("http")
+		all = all.Also(apis.ErrMissingField("http"))
+	} else {
+		if err := r.HTTP.Validate().ViaField("http"); err != nil {
+			all = all.Also(err)
+		}
```

Comment:
you can drop the `err != nil` check and just use:
```
all = all.Also(r.HTTP.Validate().ViaField("http"))
```

Reply:
Nice 👍 

Done.